### PR TITLE
Implement `tanzu builder plugin build` command as part of `builder` plugin

### PR DESCRIPTION
**Pull request by anujc25**
### What this PR does / why we need it
- This PR implements the `tanzu builder plugin build` command which builds the plugins and organizes the plugin based on the OS_ARCH grouping as well as the plugin's `target`
- The PR also updates the plugin manager to take this new artifacts directory into consideration while installing plugins with the `--local` flag.

#### Details:
`tanzu builder plugin build` can be used to build the plugins and create artifacts that can be used with tanzu cli.

The artifacts output directory structure will be different compared to what we have with `tanzu builder cli compile`
command in a way that the `tanzu builder plugin build` command will automatically group plugins based on the given os_arch
and `target`.
This command will also generate a plugin manifest file (`plugin_manifest.yaml`) that can be used in the publishing and
installation of plugins from the local directory with `tanzu plugin install --local`.

Below are the flags available with this command:

```txt
      --artifacts string      path to output artifacts directory (default "./artifacts")
  -h, --help                  help for build
      --ldflags string        ldflags to set on build
      --match string          match a plugin name to build, supports globbing (default "*")
      --os-arch stringArray   compile for specific os-arch, use 'local' for host os, use '<os>_<arch>' for specific (e.g. 'linux_amd64') (default [all])
      --path string           path of plugin directory (default "./cmd/plugin")
  -v, --version string        version of the plugins
```

Below are a few examples:

```shell
  # Build all plugins under the 'cmd/plugin' directory for the local host os and arch
  tanzu builder plugin build --path ./cmd/plugin --version v0.0.2 --os-arch local

  # Build all plugins under the 'cmd/plugin' directory for os-arch 'darwin_amd64', 'linux_amd64', 'windows_amd64'
  tanzu builder plugin build --path ./cmd/plugin --version v0.0.2 --os-arch darwin_amd64 --os-arch linux_amd64 --os-arch windows_amd64

  # Build only foo plugin under the 'cmd/plugin' directory for all supported os-arch
  tanzu builder plugin build --path ./cmd/plugin --version v0.0.2 --os-arch all --match foo
```

Here is the artifacts directory structure that gets generated by running the `tanzu builder plugin build --path ./cmd/plugin --version v0.0.2 --os-arch darwin_amd64 --os-arch linux_amd64` command.
Assuming we have 2 plugins `foo` with a `global` target and `bar` with a `kubernetes` target.

```shell
artifacts
├── darwin
│   └── amd64
│       ├── global
│       │   └── foo
│       │       └── v0.0.2
│       │           └── tanzu-foo-darwin_amd64
│       ├── kubernetes
│       │   └── bar
│       │       └── v0.0.2
│       │           └── tanzu-bar-darwin_amd64
│       └── plugin_manifest.yaml
├── linux
│   └── amd64
│       ├── global
│       │   └── foo
│       │       └── v0.0.2
│       │           └── tanzu-foo-linux_amd64
│       ├── kubernetes
│       │   └── bar
│       │       └── v0.0.2
│       │           └── tanzu-bar-linux_amd64
│       └── plugin_manifest.yaml
└── plugin_manifest.yaml

```

**Note: `tanzu builder plugin build` command expects one of 2 things:**
- Each plugin advertises the `Target` information as part of the PluginDescriptor.
- Each plugin directory contains a `metadata.yaml` file which describes the name and the target of the plugin.

**Note: With this PR there are no changes in the flags or behavior of the existing `tanzu builder cli compile` command.**

### Which issue(s) this PR fixes

Fixes #

### Describe testing done for PR
- Build the `builder` and `test` plugins within this repository with `tanzu builder plugin build` command
- Also tested the plugin installation with `tanzu plugin install all --local` by passing new artifacts directory
- Use Makefile target `make build-all` to build the core cli along with admin plugins for host os.
- Use the generated artifacts directory to install plugins `tanzu plugin install all --local artifacts-admin/darwin/adm64`

### Release note

```release-note
Implement the `tanzu builder plugin build` command as part of the `builder` plugin which builds the plugins and organizes the plugin based on the OS_ARCH grouping as well as the plugin's target
```

### CODE REVIEW

**Review by prkalle — COMMENTED** [review]
The overall changes LGTM. Thanks!

File: cmd/plugin/builder/plugin.go [hunk 1]
```diff
@@ -0,0 +1,74 @@
+// Copyright 2022 VMware, Inc. All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (anujc25) [reply cmd/plugin/builder/plugin.go:1]: Ack. I will update this as part of the next rebase.

[the remark this answers is not in the document]
Comment (anujc25) [reply cmd/plugin/builder/plugin.go]: Fixed.

File: Makefile [hunk 2]
```diff
@@ -156,16 +150,7 @@ build-plugin-admin-%: prepare-builder
...
-
-
-.PHONY: publish-admin-plugins-all-local
-publish-admin-plugins-all-local: prepare-builder ## Publish CLI admin plugins locally for all supported os-arch
-	$(BUILDER) publish --type local --plugins "$(ADMIN_PLUGINS)" --version $(BUILD_VERSION) --os-arch "${ENVS}" --local-output-discovery-dir "$(TANZU_PLUGIN_PUBLISH_PATH)/discovery/admin" --local-output-distribution-dir "$(TANZU_PLUGIN_PUBLISH_PATH)/distribution" --input-artifact-dir $(ARTIFACTS_ADMIN_DIR)
-
-.PHONY: publish-admin-plugins-local
-publish-admin-plugins-local: prepare-builder ## Publish CLI admin plugins locally for current host os-arch only
-	$(BUILDER) publish --type local --plugins "$(ADMIN_PLUGINS)" --version $(BUILD_VERSION) --os-arch "${GOHOSTOS}-${GOHOSTARCH}" --local-output-discovery-dir "$(TANZU_PLUGIN_PUBLISH_PATH)/discovery/admin" --local-output-distribution-dir "$(TANZU_PLUGIN_PUBLISH_PATH)/distribution" --input-artifact-dir $(ARTIFACTS_ADMIN_DIR)
+	$(BUILDER) plugin build --version $(BUILD_VERSION) --path ./cmd/plugin --artifacts "$(ARTIFACTS_ADMIN_DIR)" --os-arch ${OS}_${ARCH}
```
Comment (vuil) [inline Makefile:153]: do we need to retain the --ldflags passing?

File: cmd/plugin/builder/README.md [hunk 3]
```diff
@@ -35,14 +35,81 @@ Plugins will find that their `make build` command will suffice for most compile
...
+  tanzu builder plugin build --path ./cmd/plugin --version v0.0.2 --os-arch local
+
+  # Build all plugins under the 'cmd/plugin' directory for os-arch 'darwin_amd64', 'linux_amd64', 'windows_amd64'
+  tanzu builder plugin build --path ./cmd/plugin --version v0.0.2 --os-arch darwin_amd64 --os-arch linux_amd64 --os-arch windows_amd64
+
+  # Build only foo plugin under the 'cmd/plugin' directory for all supported os-arch
+  tanzu builder plugin build --path ./cmd/plugin --version v0.0.2 --os-arch all --match foo
+```
+
+Here is the artifacts directory structure that gets generated by running the `tanzu builder plugin build` command.
```
Comment (vuil) [inline cmd/plugin/builder/README.md:80]: nit: L80-81 should match the output shown or vice-versa.
+1 to showing multiple osarch and target types to illustrate the point, so I suggest showing the full `tanzu build plugin build` command (to include the --os-arch twice)

File: cmd/plugin/builder/README.md [hunk 4]
```diff
@@ -35,14 +35,81 @@ Plugins will find that their `make build` command will suffice for most compile
...
+│       ├── kubernetes
+│       │   └── bar
+│       │       └── v0.0.2
+│       │           └── tanzu-bar-linux_amd64
+│       └── plugin_manifest.yaml
+└── plugin_manifest.yaml
+
+```
+
+Note: `tanzu builder plugin build` command requires that each plugin directory contains a `metadata.yaml` file which describes the name and the target of the plugin.
```
Comment (vuil) [inline cmd/plugin/builder/README.md:111]: does the role of the metadata.yaml need to be updated?

**Review by vuil — COMMENTED** [review]
Couple of small things.

Also, 

- can you clarify in the PR description if there are any UX or output changes to the `cli compile` command?
- nit: suggest we say a little more in the release note section. Something closer to two lines what is in the `what this does...` summary of the pr description should work.

File: Makefile (hunk 2 above)
Comment (anujc25) [reply Makefile:153]: Currently, plugins in this repository doesn't need `--ldflags` and hence I have removed it. In #80 I am adding generic tooling to build any plugin with the `plugin-tooling.mk` file where I have added the `--ldflags`.

File: cmd/plugin/builder/README.md (hunk 4 above)
Comment (anujc25) [reply cmd/plugin/builder/README.md]: Updated. Thanks for catching this. I updated the PR description but forgot to update `README.md`.

File: Makefile (hunk 2 above)
Comment (vuil) [reply Makefile:153]: I do see it in #80 now, so effectively we are getting it back for plugins in this repo as well after 80 merges, so I am okay.
They are not used by default, but being able to used them is still useful, e.g. when building the plugins differently for debugging purposes.

**Review by vuil — APPROVED** [review]
lgtm, thanks for the updates:

For completeness, perhaps this should be applied to the PR description as well.

`Here is the artifacts directory structure that gets generated by running the `tanzu builder plugin build --path ./cmd/plugin --version v0.0.2 --os-arch darwin_amd64 --os-arch linux_amd64` command.`

**Pull request merged**